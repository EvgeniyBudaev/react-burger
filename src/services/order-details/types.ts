import { AxiosError } from "axios";
import { ActionTypes } from "services/order-details";
import { IOrderDetails } from "types/order";

export interface IActionOrderDetailsRequest {
    type: ActionTypes.GET_ORDER_DETAILS_REQUEST;
}

export interface IActionOrderDetailsSuccess {
    type: ActionTypes.GET_ORDER_DETAILS_SUCCESS;
    payload: IOrderDetails;
}

export interface IActionOrderDetailsFailed {
    type: ActionTypes.GET_ORDER_DETAILS_FAILED;
    payload: AxiosError;
}

export type OrderDetailsActionsType =
    | IActionOrderDetailsRequest
    | IActionOrderDetailsSuccess
    | IActionOrderDetailsFailed;
