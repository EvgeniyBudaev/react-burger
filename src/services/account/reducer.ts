import { AxiosError } from "axios";
import { Reducer } from "redux";
import { AccountActionsType, ActionTypes } from "services/account";
import { IUser } from "./types";

interface IAccountState {
    accessToken: string | null;
    refreshToken: string | null;
    emailSent: boolean;
    passwordReseted: boolean;
    success: boolean;
    user: IUser | null;
    forgotPasswordRequest: boolean;
    forgotPasswordFailed: boolean;
    loginRequest: boolean;
    loginFailed: boolean;
    registerRequest: boolean;
    registerFailed: boolean;
    resetPasswordRequest: boolean;
    resetPasswordFailed: boolean;
    error: AxiosError | null;
}

const initialState: IAccountState = {
    accessToken: null,
    refreshToken: null,
    emailSent: false,
    passwordReseted: false,
    success: false,
    user: null,
    forgotPasswordRequest: false,
    forgotPasswordFailed: false,
    loginRequest: false,
    loginFailed: false,
    registerRequest: false,
    registerFailed: false,
    resetPasswordRequest: false,
    resetPasswordFailed: false,
    error: null,
};

export const reducer: Reducer<IAccountState, AccountActionsType> = (
    state = initialState,
    action
) => {
    switch (action.type) {
        case ActionTypes.LOGIN_USER_REQUEST: {
            return {
                ...state,
                loginRequest: true,
                loginFailed: false,
                error: null,
                success: false,
            };
        }
        case ActionTypes.LOGIN_USER_SUCCESS:
            return {
                ...state,
                accessToken: action.payload.accessToken,
                refreshToken: action.payload.refreshToken,
                loginRequest: false,
                success: action.payload.success,
                user: action.payload.user,
            };
        case ActionTypes.LOGIN_USER_FAILED:
            return {
                ...state,
                loginRequest: false,
                loginFailed: true,
                error: action.payload,
            };
        case ActionTypes.REGISTER_USER_REQUEST: {
            return {
                ...state,
                registerRequest: true,
                registerFailed: false,
                error: null,
                success: false,
            };
        }
        case ActionTypes.REGISTER_USER_SUCCESS:
            return {
                ...state,
                accessToken: action.payload.accessToken,
                refreshToken: action.payload.refreshToken,
                registerRequest: false,
                success: action.payload.success,
                user: action.payload.user,
            };
        case ActionTypes.REGISTER_USER_FAILED:
            return {
                ...state,
                registerRequest: false,
                registerFailed: true,
                error: action.payload,
            };
        case ActionTypes.FORGOT_PASSWORD_REQUEST: {
            return {
                ...state,
                emailSent: false,
                forgotPasswordRequest: true,
                forgotPasswordFailed: false,
                error: null,
            };
        }
        case ActionTypes.FORGOT_PASSWORD_SUCCESS:
            return {
                ...state,
                emailSent: true,
                forgotPasswordRequest: false,
            };
        case ActionTypes.FORGOT_PASSWORD_FAILED:
            return {
                ...state,
                forgotPasswordRequest: false,
                forgotPasswordFailed: true,
                error: action.payload,
            };
        case ActionTypes.RESET_PASSWORD_REQUEST: {
            return {
                ...state,
                passwordReseted: false,
                resetPasswordRequest: true,
                resetPasswordFailed: false,
                error: null,
            };
        }
        case ActionTypes.RESET_PASSWORD_SUCCESS:
            return {
                ...state,
                passwordReseted: true,
                resetPasswordRequest: false,
            };
        case ActionTypes.RESET_PASSWORD_FAILED:
            return {
                ...state,
                resetPasswordRequest: false,
                resetPasswordFailed: true,
                error: action.payload,
            };
        default:
            return state;
    }
};
