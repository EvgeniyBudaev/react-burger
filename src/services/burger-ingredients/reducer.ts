import { Reducer } from "redux";
import { IError } from "types/error";
import { IIngredient } from "types/ingredient";
import {
    ActionTypes,
    BurgerIngredientsActionsType,
} from "services/burger-ingredients/index";

export interface IBurgerIngredientsState {
    ingredients: IIngredient[];
    ingredientsRequest: boolean;
    ingredientsFailed: boolean;
    ingredientsError: IError | null;
}

const initialState: IBurgerIngredientsState = {
    ingredients: [],
    ingredientsRequest: false,
    ingredientsFailed: false,
    ingredientsError: null,
};

export const reducer: Reducer<
    IBurgerIngredientsState,
    BurgerIngredientsActionsType
> = (state = initialState, action) => {
    switch (action.type) {
        case ActionTypes.GET_BURGER_INGREDIENTS_REQUEST: {
            return {
                ...state,
                ingredientsRequest: true,
                ingredientsFailed: false,
                ingredientsError: null,
            };
        }
        case ActionTypes.GET_BURGER_INGREDIENTS_SUCCESS:
            return {
                ...state,
                ingredients: [...action.payload].map(item => {
                    return { ...item, count: 0 };
                }),
                ingredientsRequest: false,
            };
        case ActionTypes.GET_BURGER_INGREDIENTS_FAILED:
            return {
                ...state,
                ingredientsRequest: false,
                ingredientsFailed: true,
                ingredientsError: action.payload,
            };
        default:
            return state;
    }
};
