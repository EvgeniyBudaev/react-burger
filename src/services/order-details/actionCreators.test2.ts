import fetchMock from "fetch-mock";
import configureMockStore from "redux-mock-store";
import thunk from "redux-thunk";
import { ActionTypes, fetchMakeOrder } from "services/order-details";

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

describe("order-details async actions", () => {
    afterEach(() => {
        fetchMock.restore();
    });

    it("creates GET_ORDER_DETAILS_SUCCESS", () => {
        fetchMock.postOnce("https://norma.nomoreparties.space/api/orders", {
            body: {
                data: {
                    name: "Флюоресцентный люминесцентный бургер",
                    order: {
                        createdAt: "2022-03-02T09:31:01.935Z",
                        ingredients: [
                            {
                                calories: 643,
                                carbohydrates: 85,
                                count: 0,
                                _id: "60d3b41abdacab0026a733c8",
                                id: "63d3aca1-7c2d-4f98-a932-7fd94ce4a850",
                                image: "https://code.s3.yandex.net/react/code/meat-03.png",
                                image_large:
                                    "https://code.s3.yandex.net/react/code/meat-03-large.png",
                                image_mobile:
                                    "https://code.s3.yandex.net/react/code/meat-03-mobile.png",
                                fat: 26,
                                name: "Филе Люминесцентного тетраодонтимформа",
                                price: 988,
                                proteins: 44,
                                type: "main",
                                __v: 0,
                            },
                        ],
                        name: "Флюоресцентный люминесцентный бургер",
                        number: 10911,
                        owner: {
                            createdAt: "2022-02-16T03:40:08.089Z",
                            email: "ivan@gmail.com",
                            name: "Ivan",
                            updatedAt: "2022-02-16T03:40:08.089Z",
                        },
                        price: 1976,
                        status: "done",
                        updatedAt: "2022-03-02T09:31:02.247Z",
                        _id: "621f395525b9a4001b6e1fed",
                    },
                    success: true,
                },
            },
            headers: { "content-type": "application/json" },
        });

        const expectedActions = [
            { type: ActionTypes.GET_ORDER_DETAILS_REQUEST },
            {
                type: ActionTypes.GET_ORDER_DETAILS_SUCCESS,
                payload: {
                    name: "Флюоресцентный люминесцентный бургер",
                    order: {
                        createdAt: "2022-03-02T09:31:01.935Z",
                        ingredients: [
                            {
                                calories: 643,
                                carbohydrates: 85,
                                count: 0,
                                _id: "60d3b41abdacab0026a733c8",
                                id: "63d3aca1-7c2d-4f98-a932-7fd94ce4a850",
                                image: "https://code.s3.yandex.net/react/code/meat-03.png",
                                image_large:
                                    "https://code.s3.yandex.net/react/code/meat-03-large.png",
                                image_mobile:
                                    "https://code.s3.yandex.net/react/code/meat-03-mobile.png",
                                fat: 26,
                                name: "Филе Люминесцентного тетраодонтимформа",
                                price: 988,
                                proteins: 44,
                                type: "main",
                                __v: 0,
                            },
                        ],
                        name: "Флюоресцентный люминесцентный бургер",
                        number: 10911,
                        owner: {
                            createdAt: "2022-02-16T03:40:08.089Z",
                            email: "ivan@gmail.com",
                            name: "Ivan",
                            updatedAt: "2022-02-16T03:40:08.089Z",
                        },
                        price: 1976,
                        status: "done",
                        updatedAt: "2022-03-02T09:31:02.247Z",
                        _id: "621f395525b9a4001b6e1fed",
                    },
                    success: true,
                },
            },
            { type: ActionTypes.CLEAR_ALL_INGREDIENTS },
            {
                type: ActionTypes.GET_ORDER_DETAILS_FAILED,
                payload: {
                    error: {
                        body: "Не правильные параметры запроса!",
                        message: "Request failed with status code 403",
                    },
                    success: false,
                },
            },
        ];
        const store = mockStore({
            details: {
                name: "",
                order: {
                    createdAt: "",
                    _id: "",
                    ingredients: [],
                    name: "",
                    number: undefined,
                    owner: {
                        createdAt: "",
                        email: "",
                        name: "",
                        updatedAt: "",
                    },
                    status: undefined,
                    updatedAt: "",
                },
                success: false,
            },
            detailsRequest: false,
            detailsFailed: false,
            detailsError: null,
        });

        return store.dispatch(fetchMakeOrder()).then(() => {
            expect(store.getActions()).toEqual(expectedActions);
        });
    });
});
