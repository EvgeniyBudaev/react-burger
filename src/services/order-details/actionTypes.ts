export enum ActionTypes {
    CLEAR_ALL_INGREDIENTS = "CLEAR_ALL_INGREDIENTS",
    GET_ORDER_DETAILS_REQUEST = "GET_ORDER_DETAILS_REQUEST",
    GET_ORDER_DETAILS_SUCCESS = "GET_ORDER_DETAILS_SUCCESS",
    GET_ORDER_DETAILS_FAILED = "GET_ORDER_DETAILS_FAILED",
}
