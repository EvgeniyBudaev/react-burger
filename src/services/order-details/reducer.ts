import { Reducer } from "redux";
import { ActionTypes, OrderDetailsActionsType } from "services/order-details";
import { IOrderDetails } from "types/order";
import { IError } from "types/error";

export interface IOrderDetailsState {
    details: IOrderDetails;
    detailsRequest: boolean;
    detailsFailed: boolean;
    detailsError: IError | null;
}

const initialState: IOrderDetailsState = {
    details: {
        name: "",
        order: {
            createdAt: "",
            _id: "",
            ingredients: [],
            name: "",
            number: undefined,
            owner: {
                createdAt: "",
                email: "",
                name: "",
                updatedAt: "",
            },
            status: undefined,
            updatedAt: "",
        },
        success: false,
    },
    detailsRequest: false,
    detailsFailed: false,
    detailsError: null,
};

export const reducer: Reducer<IOrderDetailsState, OrderDetailsActionsType> = (
    state = initialState,
    action
) => {
    switch (action.type) {
        case ActionTypes.GET_ORDER_DETAILS_REQUEST: {
            return {
                ...state,
                detailsRequest: true,
                detailsFailed: false,
                detailsError: null,
            };
        }
        case ActionTypes.GET_ORDER_DETAILS_SUCCESS:
            return {
                ...state,
                details: action.payload,
                detailsRequest: false,
            };
        case ActionTypes.GET_ORDER_DETAILS_FAILED:
            return {
                ...state,
                detailsRequest: false,
                detailsFailed: true,
                detailsError: action.payload,
            };
        default:
            return state;
    }
};
