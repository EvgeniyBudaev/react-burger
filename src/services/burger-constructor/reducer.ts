import { INGREDIENT_TYPE } from "constants/ingredient";
import { Reducer } from "redux";
import update from "immutability-helper";
import { IIngredient } from "types/ingredient";
import { ActionTypes } from "./actionTypes";
import { BurgerConstructorActionsType } from "./types";

interface IBurgerConstructorState {
    bun: IIngredient | Record<string, never>;
    mains: IIngredient[];
}

const initialState: IBurgerConstructorState = {
    bun: {},
    mains: [],
};

export const reducer: Reducer<
    IBurgerConstructorState,
    BurgerConstructorActionsType
> = (state = initialState, action) => {
    switch (action.type) {
        case ActionTypes.ADD_INGREDIENT: {
            if (action.payload.type === INGREDIENT_TYPE.BUN) {
                return {
                    ...state,
                    bun: action.payload,
                };
            } else
                return {
                    ...state,
                    mains: [...state.mains, action.payload],
                };
        }
        case ActionTypes.DELETE_INGREDIENT: {
            return {
                ...state,
                mains: update(state.mains, {
                    $splice: [[action.payload, 1]],
                }),
            };
        }
        case ActionTypes.MOVE_INGREDIENTS: {
            const draggingIngredient = state.mains[action.dragIndex];
            return {
                ...state,
                mains: update(state.mains, {
                    $splice: [
                        [action.dragIndex, 1],
                        [action.hoverIndex, 0, draggingIngredient],
                    ],
                }),
            };
        }
        default:
            return state;
    }
};
