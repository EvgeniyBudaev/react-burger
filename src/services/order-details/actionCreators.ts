import { BASE_URL } from "constants/routes";
import axios, { AxiosError } from "axios";
import { ActionTypes } from "services/order-details";
import { ActionTypes as ActionTypesBurgerConstructor } from "services/burger-constructor";
import { AppDispatch, AppThunk } from "services/types";
import { IOrderDetails } from "types/order";
import { getCookie } from "utils/coockie";
import { getErrorByStatus } from "utils/error";

export interface IOrderDetailsRequest {
    ingredients: string[];
}

const ORDER_DETAILS_URL = `${BASE_URL}orders`;
const headers = {
    Accept: "application/json",
    "Content-Type": "application/json",
};

export const fetchMakeOrder: AppThunk =
    (options: IOrderDetailsRequest) => async (dispatch: AppDispatch) => {
        try {
            const config = {
                headers: {
                    ...headers,
                    authorization: `Bearer ${getCookie("accessToken")}`,
                },
            };
            const body = JSON.stringify(options);
            dispatch({ type: ActionTypes.GET_ORDER_DETAILS_REQUEST });
            const response = await axios.post<IOrderDetails>(
                ORDER_DETAILS_URL,
                body,
                config
            );
            console.log("response:", response);
            dispatch({
                type: ActionTypes.GET_ORDER_DETAILS_SUCCESS,
                payload: response.data,
            });
            dispatch({
                type: ActionTypesBurgerConstructor.CLEAR_ALL_INGREDIENTS,
            });
        } catch (e) {
            const error = e as AxiosError;
            const errorByStatus = getErrorByStatus(error);
            dispatch({
                type: ActionTypes.GET_ORDER_DETAILS_FAILED,
                payload: errorByStatus,
            });
        }
    };
