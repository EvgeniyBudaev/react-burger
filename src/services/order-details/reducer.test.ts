import { ActionTypes, IOrderDetailsState } from "services/order-details";
import { reducer } from "./reducer";

const initialState: IOrderDetailsState = {
    details: {
        name: "",
        order: {
            createdAt: "",
            _id: "",
            ingredients: [],
            name: "",
            number: undefined,
            owner: {
                createdAt: "",
                email: "",
                name: "",
                updatedAt: "",
            },
            status: undefined,
            updatedAt: "",
        },
        success: false,
    },
    detailsRequest: false,
    detailsFailed: false,
    detailsError: null,
};

describe("order-details reducer", () => {
    it("should return the initial state", () => {
        expect(initialState).toMatchSnapshot();
    });

    it("should handle GET_ORDER_DETAILS_REQUEST", () => {
        expect(
            reducer(initialState, {
                type: ActionTypes.GET_ORDER_DETAILS_REQUEST,
            })
        ).toEqual(
            expect.objectContaining({
                ...initialState,
                detailsRequest: true,
                detailsFailed: false,
                detailsError: null,
            })
        );
    });

    it("should handle GET_ORDER_DETAILS_SUCCESS", () => {
        expect(
            reducer(initialState, {
                type: ActionTypes.GET_ORDER_DETAILS_SUCCESS,
                payload: {
                    name: "Флюоресцентный люминесцентный бургер",
                    success: true,
                    order: {
                        ingredients: [
                            "60d3b41abdacab0026a733c8",
                            "60d3b41abdacab0026a733c6",
                        ],
                        _id: "621f05d225b9a4001b6e1fb1",
                        owner: {
                            name: "Ivan",
                            email: "ivan@gmail.com",
                            createdAt: "2022-02-16T03:40:08.089Z",
                            updatedAt: "2022-02-16T03:40:08.089Z",
                        },
                        status: "done",
                        name: "Флюоресцентный люминесцентный бургер",
                        createdAt: "2022-03-02T05:51:14.872Z",
                        updatedAt: "2022-03-02T05:51:15.211Z",
                        number: 10908,
                        price: 1976,
                    },
                },
            })
        ).toEqual(
            expect.objectContaining({
                ...initialState,
                detailsRequest: false,
                details: {
                    name: "Флюоресцентный люминесцентный бургер",
                    success: true,
                    order: {
                        ingredients: [
                            "60d3b41abdacab0026a733c8",
                            "60d3b41abdacab0026a733c6",
                        ],
                        _id: "621f05d225b9a4001b6e1fb1",
                        owner: {
                            name: "Ivan",
                            email: "ivan@gmail.com",
                            createdAt: "2022-02-16T03:40:08.089Z",
                            updatedAt: "2022-02-16T03:40:08.089Z",
                        },
                        status: "done",
                        name: "Флюоресцентный люминесцентный бургер",
                        createdAt: "2022-03-02T05:51:14.872Z",
                        updatedAt: "2022-03-02T05:51:15.211Z",
                        number: 10908,
                        price: 1976,
                    },
                },
            })
        );
    });

    it("should handle GET_ORDER_DETAILS_FAILED", () => {
        expect(
            reducer(initialState, {
                type: ActionTypes.GET_ORDER_DETAILS_FAILED,
                payload: {
                    success: false,
                    error: {
                        body: "error body",
                        message: "error message",
                    },
                },
            })
        ).toEqual(
            expect.objectContaining({
                ...initialState,
                detailsRequest: false,
                detailsFailed: true,
                detailsError: {
                    success: false,
                    error: {
                        body: "error body",
                        message: "error message",
                    },
                },
            })
        );
    });
});
