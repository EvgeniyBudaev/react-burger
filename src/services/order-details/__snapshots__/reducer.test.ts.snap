// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`order-details reducer should return the initial state 1`] = `
Object {
  "details": Object {
    "name": "",
    "order": Object {
      "_id": "",
      "createdAt": "",
      "ingredients": Array [],
      "name": "",
      "number": undefined,
      "owner": Object {
        "createdAt": "",
        "email": "",
        "name": "",
        "updatedAt": "",
      },
      "status": undefined,
      "updatedAt": "",
    },
    "success": false,
  },
  "detailsError": null,
  "detailsFailed": false,
  "detailsRequest": false,
}
`;
